scylla-cdc-replicator/src/replicator_consumer.rs

```diff
@@ -0,0 +1,248 @@
+            values_for_update[1] = value;
+            self.run_statement(
+                Query::new(format!(
+                    "UPDATE {}.{} USING TTL ? SET {} = ? WHERE {}",
+                    self.dest_keyspace_name, self.dest_table_name, column_name, self.keys_cond
+                )),
+                values_for_update,
+                timestamp,
+            )
+            .await;
+        } else if data.is_value_deleted(column_name) {
+            self.run_statement(
+                Query::new(format!(
+                    "DELETE {} FROM {}.{} WHERE {}",
+                    column_name, self.dest_keyspace_name, self.dest_table_name, self.keys_cond
+                )),
+                values_for_delete,
+                timestamp,
+            )
+            .await;
+        }
+    }
+
+    async fn update_or_insert(&self, data: CDCRow<'_>, is_insert: bool) {
+        let keys_iter = ReplicatorConsumer::get_keys_iter(&self.table_schema);
+        let ttl = CqlValue::Int(data.ttl.unwrap_or(0) as i32); // If data is inserted without TTL, setting it to 0 deletes existing TTL.
+        let timestamp = ReplicatorConsumer::get_timestamp(&data);
+        let values = keys_iter
+            .clone()
+            .map(|col_name| data.get_value(col_name).as_ref().unwrap())
+            .collect::<Vec<&CqlValue>>();
+
+        if is_insert {
+            // Insert row with nulls, the rest will be done through an update.
+            let mut insert_values = Vec::with_capacity(values.len() + 1);
+            insert_values.extend(values.iter());
+            insert_values.push(&ttl);
+
+            self.run_prepared_statement(self.insert_query.clone(), &insert_values, timestamp)
+                .await;
+        }
+
+        let mut values_for_update = Vec::with_capacity(2 + values.len());
+        values_for_update.extend([&ttl, &CqlValue::Int(0)]);
+        values_for_update.extend(values.iter());
+
+        for column_name in &self.non_key_columns {
+            match self.table_schema.columns.get(column_name).unwrap().type_ {
+                CqlType::Native(_)
+                | CqlType::Tuple(_)
+                | CqlType::Collection { frozen: true, .. }
+                | CqlType::UserDefinedType { frozen: true, .. } => {
+                    self.update_atomic_or_frozen(
+                        column_name,
+                        &data,
+                        &mut values_for_update,
+                        &values,
+                        timestamp,
+                    )
+                    .await
+                }
+                _ => todo!("This type of data can't be replicated yet!"),
+            }
+        }
+    }
+
+    async fn run_prepared_statement(
+        &self,
+        mut query: PreparedStatement,
+        values: &[&CqlValue],
+        timestamp: i64,
+    ) {
+        query.set_timestamp(Some(timestamp));
+        self.dest_session
+            .execute(&query, values)
```

Comment:
`execute` and `query` API should probably be extended to take a timestamp param so you don't have to clone the `query` every time